File: day12/src/main/cpp/solution.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <string>
#include <vector>

using namespace std;

struct entry {
  string node;
  set<string> seen;
  bool doubled;
};

map<string, vector<string>> readInputFile() {
  ifstream file("../resources/input");
  map<string, vector<string>> routes;
  string s, from, to;
  int delim;
  while (getline(file, s)) {
    delim = s.find("-");
    from = s.substr(0, delim);
    to = s.substr(delim + 1, s.length());
    routes.emplace(from, vector<string>());
    routes.emplace(to, vector<string>());
    routes[from].push_back(to);
    routes[to].push_back(from);
  }
  return routes;
}

int countPossibleRoutes(map<string, vector<string>> &routes,
                        bool doubleRule = false) {
  queue<entry> q;
  q.push({"start", set<string>(), false});
  int count = 0;
  while (!q.empty()) {
    auto current = q.front();
    q.pop();
    for (auto const &option : routes[current.node]) {
      if (option == "end") {
        ++count;
      } else if (option == "start") {
        continue;
      } else if (option[0] >= 'a' && option[0] <= 'z') {
        if (current.seen.find(option) == current.seen.end()) {
          auto newSeen = current.seen;
          newSeen.insert(option);
          q.push({option, newSeen, current.doubled});
        } else if (doubleRule && !current.doubled) {
          q.push({option, current.seen, true});
        }
      } else {
        q.push({option, current.seen, current.doubled});
      }
    }
  }
  return count;
}

int main() {
  auto routes = readInputFile();
  cout << "part 1" << endl
       << countPossibleRoutes(routes) << endl
       << "part 2" << endl
       << countPossibleRoutes(routes, true) << endl;
  return 0;
}

File: day12/src/main/python/solution.py
def read_input():
    f = open('../resources/input', 'r')
    routes = dict()
    for line in f.readlines():
        start, end = line.strip().split("-")
        if start not in routes.keys():
            routes[start] = []
        if end not in routes.keys():
            routes[end] = []
        routes[start].append(end)
        routes[end].append(start)
    return routes


def find_possible_routes(routes, double=False):
    queue = [["start", {'start'}, False]]
    count = 0
    while len(queue) != 0:
        node, original, doubled = queue.pop()
        for target in routes[node]:
            seen = original.copy()
            if target == "end":
                count += 1
            elif target == "start":
                continue
            elif target.islower():
                if target not in seen:
                    seen.add(target)
                    queue.append([target, seen, doubled])
                elif double and not doubled:
                    queue.append([target, seen, True])
            else:
                queue.append([target, seen, doubled])
    return count


if __name__ == "__main__":
    route_list = read_input()
    print("part 1")
    print(find_possible_routes(route_list))
    print("part 2")
    print(find_possible_routes(route_list, True))
